L1Trigger/L1THGCal/plugins/fe_codecs/HGCalBestChoiceCodec.cc

```diff
@@ -0,0 +1,182 @@
+                if(data[index]) value |= (0x1<<i);
+            }
+            c++;
+            result.payload[b] = value;
+        }
+    }
+    return result;
+}
+
+
+/*****************************************************************/
+void HGCalBestChoiceCodec::triggerCellSums(const HGCalTriggerGeometryBase& geom, const std::vector<HGCEEDataFrame>& dataframes)
+/*****************************************************************/
+{
+    std::map<HGCTriggerDetId, uint32_t> payload;
+    // sum energies in trigger cells
+    for(const auto& frame : dataframes)
+    {
+        // FIXME: only EE
+        HGCEEDetId cellid(frame.id());
+        HGCTriggerDetId triggercellid( geom.getTriggerCellFromCell(cellid)->triggerCellId() );
+        payload.insert( std::make_pair(triggercellid, 0) ); // do nothing if key exists already
+        // FIXME: need to transform ADC and TDC to the same linear scale on 12 bits
+        uint32_t data = frame[2].data(); // 'data' has to be a 12 bit word
+        payload[triggercellid] += data; // 32 bits integer should be largely enough (maximum 7 12-bits sums are done)
+
+    }
+    // fill data payload
+    for(const auto& id_value : payload)
+    {
+        uint32_t id = id_value.first.cell();
+        if(id>nCellsInModule_) // cell number starts at 1
+        {
+            edm::LogWarning("HGCalBestChoiceCodec") 
+                << "Number of trigger cells in module too large for available data payload\n";
+            continue;
+        }
+        data_.payload.at(id-1) = id_value.second;
+    }
+}
+
+
+/*****************************************************************/
+void HGCalBestChoiceCodec::bestChoiceSelect()
+/*****************************************************************/
+{
+    // Store data payload in vector for energy sorting. Then refill the data payload after trigger
+    // cell selection.
+    // Probably not the most efficient way.
+    // Should check in the firmware how cells with the same energy are sorted
+
+    // copy for sorting
+    std::vector< std::pair<uint32_t, uint32_t> > sortedtriggercells; // value, ID
+    sortedtriggercells.reserve(nCellsInModule_);
+    for(size_t i=0; i<nCellsInModule_; i++)
+    {
+        sortedtriggercells.push_back(std::make_pair(data_.payload[i], i));
+    }
+    // sort, reverse order
+    sort(sortedtriggercells.begin(), sortedtriggercells.end(),
+            [](const std::pair<uint32_t, uint32_t>& a, 
+                const  std::pair<uint32_t, uint32_t>& b) -> bool
+            { 
+                return a > b; 
+            } 
+            );
+    // keep only the 12 first trigger cells
+    for(size_t i=nData_; i<nCellsInModule_; i++)
+    {
+        sortedtriggercells.at(i).first = 0;
+    }
+    for(const auto& value_id : sortedtriggercells)
+    {
+        if(value_id.second>nCellsInModule_) // cell number starts at 1
+        {
+            edm::LogWarning("HGCalBestChoiceCodec") 
```

Comment:
This should be an exception, it represents a badly configured state of the hardware we are simulating.